test/cpp/api/dataloader.cpp

```diff
@@ -1274,3 +1277,283 @@ TEST(DataLoaderTest, StatefulDatasetWithCollate) {
+          double,
+          samplers::RandomSampler,
+          samplers::RandomSampler>>(config);
+  auto data_loader = torch::data::make_chunk_data_loader(
+      shared_dataset,
+      DataLoaderOptions()
+          .workers(total_worker)
+          .max_jobs(max_jobs)
+          .batch_size(batch_size)
+          .chunk_loading(true));
+
+  shared_dataset->reset();
+  auto iterator = data_loader->begin();
+  size_t count_example = 0;
+  // TODO: Because current DataLoader can return empty batches,
+  // batch max_jobs to ensure chunk is fully read.
+  // Empty batches are ignored by the tests
+  for (size_t i = 0; i < max_jobs; ++i, ++iterator) {
+    std::vector<ctf::CTFExample<double>> batch = *iterator;
+    if (batch.size() != 0) {
+      ASSERT_EQ(batch.size(), batch_size);
+      ASSERT_EQ(batch[i].features[0].input_name, "word");
+      ASSERT_EQ(batch[i].labels[0].input_name, "tag");
+      count_example += batch.size();
+    }
+  }
+  ASSERT_EQ(total_example, count_example);
+}
+
+// ctf_sample_multiple_chunks_0000[0...6].ctf has 7 chunks with 3 batches each
+// (last chunk has 1 batch) and one example per batch
+TEST(DataTest, CTFDataLoaderWithChunkSupportMultipleWorkersMultipleChunks) {
+  const size_t batch_size = 3;
+  const size_t total_example = 7;
+  const size_t total_prefetch = 2;
+  const size_t total_worker = 10;
+  const size_t max_jobs = 2 * total_worker;
+
+  std::vector<torch::data::ctf::CTFConfigHelper> configs;
+  std::vector<torch::data::ctf::CTFStreamInformation> features_info;
+  std::vector<torch::data::ctf::CTFStreamInformation> labels_info;
+  features_info.emplace_back(
+      "word", "word", 0, torch::data::ctf::CTFValueFormat::Sparse);
+  labels_info.emplace_back(
+      "tag", "tag", 0, torch::data::ctf::CTFValueFormat::Sparse);
+
+  for (size_t i = 0; i < total_example; ++i) {
+    torch::data::ctf::CTFConfigHelper config(
+        std::string(
+            torch::data::ctf::CTF_SAMPLE_DIR +
+            "/ctf_sample_multiple_chunks_000" + std::to_string(i) + ".ctf"),
+        features_info,
+	labels_info,
+        torch::data::ctf::CTFDataType(torch::data::ctf::CTFDataType::Double));
+
+    configs.push_back(config);
+  }
+
+  datasets::SharedBatchDataset<ctf::CTFChunkDataset<
+      double,
+      samplers::RandomSampler,
+      samplers::RandomSampler>>
+      shared_dataset = datasets::make_shared_dataset<ctf::CTFChunkDataset<
+          double,
+          samplers::RandomSampler,
+          samplers::RandomSampler>>(configs, total_prefetch);
+  auto data_loader = torch::data::make_chunk_data_loader(
+      shared_dataset,
+      DataLoaderOptions()
+          .workers(total_worker)
+          .max_jobs(max_jobs)
+          .batch_size(batch_size)
+          .chunk_loading(true));
+
+  shared_dataset->reset();
+  auto iterator = data_loader->begin();
+  size_t count_example = 0;
+  // TODO: Because current DataLoader can return empty batches,
```

Comment:
This should not be the case with the new implementation of the chunkdataset. Could you please check?

Reply:
I will review/integrate with https://github.com/xzhu1900/pytorch/pull/11/files and update this PR